### PR TITLE
fix(engine-core): fix subclass not inheriting `@api` in dev mode

### LINKED ISSUE #3758: In dev mode, public APIs extending from superclasses may throw when called
### Description

LWC 3.4.0 seems to have introduced a regression (possibly caused by #3713) in the following case:

```js
// superclass.js
export default class extends LightningElement {
  @api
  foo() {}
}
```

```js
// subclass.js
export default class extends Superclass {
  foo() {}
}
```

(Note that the superclass has an `@api`, but the subclass _does not_.)

Before v3.4.0, a consumer of the subclass could call e.g.:

```js
elm.foo()
```

... and it would call the method correctly. In 3.4.0, however, an error is thrown:

    Uncaught TypeError: elm.foo is not a function

- [Repro in 3.4.0](https://stackblitz.com/edit/salesforce-lwc-qral4p?file=src%2Fmodules%2Fx%2Fsuperclass%2Fsuperclass.js,src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fsubclass%2Fsubclass.js,package.json&title=LWC%20playground) (throws error)
- [Comparison repro in 3.3.3](https://stackblitz.com/edit/salesforce-lwc-dt1dhz?file=src%2Fmain.js,src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fapp%2Fapp.js,src%2Fmodules%2Fx%2Fsubclass%2Fsubclass.js,rollup.config.js&title=LWC%20playground) (does not throw error)

Note that production mode is unaffected. This only affects dev mode.

### PR DESCRIPTION
**Pull request by nolanlawson**
## Details

Fixes #3758

For now I am just disabling the helpful dev warnings if a component does not directly extend from `LightningElement`. I opened some other issues to track follow-up work:

- #3761
- #3762

## Does this pull request introduce a breaking change?

* ✅ No, it does not introduce a breaking change.

## Does this pull request introduce an observable change?

* ✅ No, it does not introduce an observable change.

## GUS work item
W-14208047

### CODE REVIEW

File: packages/@lwc/integration-karma/test/component/api-with-superclasses/x/subClass/subClass.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
+import { api } from 'lwc';
```
Comment (nolanlawson) [inline packages/@lwc/integration-karma/test/component/api-with-superclasses/x/subClass/subClass.js:1]: These files are all generated by a script fwiw

File: packages/@lwc/integration-karma/test/component/api-with-superclasses/index.spec.js [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+                        ].includes('public');
+                        // TODO [#3761]: components that don't extend LightningElement should warn for missing @api
+                        // These cases are very inconsistent, but this is the current behavior
+                        const expectWarn =
+                            (subClassVariant === 'omit' &&
+                                superSuperClassVariant === 'private' &&
+                                superClassVariant !== 'public') ||
+                            (subClassVariant === 'private' &&
+                                superClassVariant !== 'public' &&
+                                superSuperClassVariant === 'private');
```
Comment (jmsjtu) [inline packages/@lwc/integration-karma/test/component/api-with-superclasses/index.spec.js:70]:
```suggestion
                        const expectWarn =
                            subClassVariant !== 'public' &&
                            superClassVariant !== 'public' &&
                            superSuperClassVariant === 'private';
```

File: packages/@lwc/integration-karma/test/component/api-with-superclasses/index.spec.js [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                    it('call on subClass', () => {
+                        // TODO [#3762]: components should not "inherit" @api from their superclasses
+                        const expectUndefined = ![
+                            subClassVariant,
+                            superClassVariant,
+                            superSuperClassVariant,
+                        ].includes('public');
+                        // TODO [#3761]: components that don't extend LightningElement should warn for missing @api
+                        // These cases are very inconsistent, but this is the current behavior
+                        const expectWarn =
```
Comment (jmsjtu) [inline packages/@lwc/integration-karma/test/component/api-with-superclasses/index.spec.js:64]: Might be worth adding a comment here to explain when a warning is expected.

Something like "The warning accessor is only placed on the `SuperSuperClass` and is invoked for each call to the private methods".

**Review by jmsjtu — APPROVED** [review]
I added some small nitpicks but LGTM!

**Pull request merged**